crates/qasm3/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+
+use oq3_semantics::syntax_to_semantics::parse_source_string;
+
+use crate::error::QASM3ImporterError;
+
+/// The name of a Python attribute to define on the given module where the default implementation
+/// of the ``stdgates.inc`` custom instructions is located.
+const STDGATES_INC_CUSTOM_GATES_ATTR: &str = "STDGATES_INC_GATES";
+
+/// Load an OpenQASM 3 program from a string into a :class:`.QuantumCircuit`.
+///
+/// Args:
+///     source (str): the program source in a Python string.
+///     custom_gates (Iterable[CustomGate]): Python constructors to use for particular named gates.
+///         If not supplied, Qiskit will use its own standard-library constructors for gates
+///         defined in the OpenQASM 3.0 standard-library file ``stdgates.inc``.
+///     include_path (Iterable[str]): the path to search when resolving ``include`` statements.
+///         If not given, Qiskit will arrange for this to point to a location containing
+///         ``stdgates.inc`` only.  Paths are tried in the sequence order.
+///
+/// Returns:
+///     :class:`.QuantumCircuit`: the constructed circuit object.
+///
+/// Raises:
+///     :class:`.QASM3ImporterError`: if an error occurred during parsing or semantic analysis.
+///         In the case of a parsing error, most of the error messages are printed to the terminal
+///         and formatted, for better legibility.
+#[pyfunction]
+#[pyo3(pass_module, signature = (source, /, *, custom_gates=None, include_path=None))]
+pub fn loads(
+    module: &PyModule,
+    py: Python,
+    source: String,
+    custom_gates: Option<Vec<circuit::PyGate>>,
+    include_path: Option<Vec<OsString>>,
+) -> PyResult<circuit::PyCircuit> {
+    let default_include_path = || -> PyResult<Vec<OsString>> {
+        Ok(vec![Path::new(module.filename()?)
+            .parent()
+            .unwrap()
+            .join(["qasm", "libs", "dummy"].iter().collect::<PathBuf>())
+            .into_os_string()])
+    };
+    let include_path = include_path.map(Ok).unwrap_or_else(default_include_path)?;
+    let result = parse_source_string(source, None, Some(&include_path));
+    if result.any_errors() {
+        result.print_errors();
+        return Err(QASM3ImporterError::new_err(
+            "errors during parsing; see printed errors",
+        ));
+    }
+    let gates = match custom_gates {
+        Some(gates) => gates
+            .into_iter()
+            .map(|gate| (gate.name().to_owned(), gate))
+            .collect(),
+        None => module
+            .getattr(STDGATES_INC_CUSTOM_GATES_ATTR)?
+            .iter()?
+            .map(|obj| {
+                let gate = obj?.extract::<circuit::PyGate>()?;
+                Ok((gate.name().to_owned(), gate))
+            })
+            .collect::<PyResult<HashMap<_, _>>>()?,
+    };
+    crate::build::convert_asg(py, result.program(), result.symbol_table(), gates)
+}
+
+/// Load an OpenQASM 3 program from a source file into a :class:`.QuantumCircuit`.
```

Comment:
This needs to have a warning about it being an experimental unstable interface.

Reply:
Done in 6e729bfdd7.